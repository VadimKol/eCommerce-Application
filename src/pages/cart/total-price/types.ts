import type { Cart, LineItem } from '@commercetools/platform-sdk';

import type { Promocode } from '@/contexts/promocodes-context';

export interface TotalPriceProps {
  addPromocode: (code: string) => Promise<void>;
  promocodes: Promocode[];
  cartItems: LineItem[];
  cart: Cart;
}
