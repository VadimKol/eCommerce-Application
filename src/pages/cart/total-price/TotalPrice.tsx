import { CustomButton } from '@/components/custom-button/Ð¡ustomButton';
import { Promocodes } from '@/components/promocodes/Promocodes';

import styles from './styles.module.scss';
import type { TotalPriceProps } from './types';

export function TotalPrice({ addPromocode, promocodes, cartItems, cart }: TotalPriceProps): JSX.Element {
  const total = (cart.totalPrice.centAmount / 100).toFixed(2);
  const subtotal = (
    cartItems.reduce((acc, item) => acc + item.price.value.centAmount * item.quantity, 0) / 100
  ).toFixed(2);
  const discount = (Number(subtotal) - Number(total)).toFixed(2);

  const isDiscounted =
    cartItems.some((item) => item.price.discounted !== undefined) ||
    cartItems.some((item) => item.discountedPricePerQuantity.length > 0) ||
    cart.discountOnTotalPrice;

  return (
    <div className={styles.total_price_block}>
      <p className={styles.total_price}>
        <span>Total:</span>
        <span className={styles.price}>${total}</span>
      </p>
      {isDiscounted && (
        <>
          <p className={styles.discount_block}>
            <span>Discount:</span>
            <span className={styles.price}>${discount}</span>
          </p>
          <p className={styles.original_price_block}>
            <span>Subtotal:</span>
            <span className={styles.discount}>${subtotal}</span>
          </p>
        </>
      )}
      <Promocodes promocodes={promocodes} addPromocode={addPromocode} cart={cart} />
      <CustomButton type="button" aria-label="Check out" className={styles.check_out} variant="tertiary">
        Check out
      </CustomButton>
    </div>
  );
}
