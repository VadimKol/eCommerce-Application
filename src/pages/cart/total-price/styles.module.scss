.total_price_block {
  flex-shrink: 0;
  width: 400px;
  height: max-content;
  padding: 30px;
  background: linear-gradient(var(--background-color) 0 0) padding-box, linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;


  @media (width <= 1300px) {
    width: 320px;
    padding: 20px;
  }

  @media (width <= 950px) {
    display: flex;
    flex-direction: column;
    width: 80%;
  }

  @media (width <= 450px) {
    width: 100%;
  }
}

.total_price,
.discount_block,
.original_price_block {
  display: flex;
  justify-content: space-between;
}

.total_price {
  margin: 0;
  font-size: 2rem;

  @media (width <= 1300px) {
    font-size: 1.4rem;
  }
}

.price,
.discount {
  font-weight: 600;
}

.discount {
  text-decoration: line-through;
  color: var(--error-color);
}

.discount_block,
.original_price_block {
  margin: 0;
  font-size: 1.5rem;

  @media (width <= 1300px) {
    font-size: 1rem;
  }
}

.original_price_block {
  margin-bottom: 0.66em;
}

.check_out {
  width: 100%;
  font-size: 1.3rem;

  @media (width <= 1300px) {
    font-size: 1rem;
  }

  @media (width <= 950px) {
    width: 240px;
    margin: auto;
    font-size: 1.3rem;
  }
}
