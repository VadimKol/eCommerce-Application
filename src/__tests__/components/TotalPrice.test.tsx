import type { Cart } from '@commercetools/platform-sdk';
import { render } from '@testing-library/react';

import { TotalPrice } from '@/pages/cart/total-price/TotalPrice';

it('renders correctly', () => {
  const cart: Cart = {
    id: '',
    version: 0,
    lineItems: [],
    customLineItems: [],
    totalPrice: {
      type: 'centPrecision',
      centAmount: 0,
      currencyCode: 'USD',
      fractionDigits: 0,
    },
    taxMode: 'Platform',
    taxRoundingMode: 'HalfEven',
    taxCalculationMode: 'LineItemLevel',
    inventoryMode: 'None',
    cartState: 'Active',
    shippingMode: 'Single',
    shipping: [],
    itemShippingAddresses: [],
    discountCodes: [],
    directDiscounts: [],
    refusedGifts: [],
    origin: 'Customer',
    createdAt: '',
    lastModifiedAt: '',
  };
  const { container } = render(
    <TotalPrice
      addPromocode={jest.fn()}
      promocodes={[{ id: '123456', name: 'GEEK-SHOP' }]}
      cartItems={[]}
      cart={cart}
    />,
  );
  expect(container).toMatchSnapshot();
});
