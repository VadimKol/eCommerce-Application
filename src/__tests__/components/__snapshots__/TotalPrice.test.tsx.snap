// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="total_price_block"
  >
    <p
      class="total_price"
    >
      <span>
        Total:
      </span>
      <span
        class="price"
      >
        $
        0.00
      </span>
    </p>
    <form
      class="promobox"
    >
      <p
        class="promo_desc"
      >
        Enter promocode
      </p>
      <div
        class="promoblock"
      >
        <input
          class="promocode"
          id="promocode"
          placeholder="Promocode"
          type="text"
        />
        <button
          class="button button--primary"
          id=""
          type="submit"
        >
          Apply
        </button>
      </div>
    </form>
    <button
      class="button button--tertiary check_out"
      id=""
      type="button"
    >
      Check out
    </button>
  </div>
</div>
`;
